### ISSUE TITLE
Dumping metadata from memory?

### ISSUE THREAD

**Issue by pauls1024** [lizardfs/lizardfs#618]
Hi I'd like to get a copy / dump / backup of the metadata from memory in a running lizardfs master... the problem is that the filesystem has gone read-only.

The filesystem is currently running and mounted by clients, so I have an up to date rsync copy of the whole filesystem.  Worst case I can build a new lizardfs master and chunkservers and then rsync the files back on to it... So I'm not going to loose any data.

Unfortunately my lizard shadow metadataserver and metalogger are now out of sync too due to a bit of an error when trying to resync things ... so lets ignore them.

I'm running Debian Jessie with the following on the master:
root@lizardfs-master:/# dpkg -l | grep lizard
ii  lizardfs-common            3.11.3                   amd64        LizardFS common files
ii  lizardfs-master            3.11.3                   amd64        LizardFS master server

Trying to use mfsmetadump to dump the current metadata to a tmpfs mount point gives me the following output, so I assume it is trying to read  /var/lib/mfs/metadata.mfs rather than generating it from the in-memory copy?
```
root@lizardfs-master:/# mfsmetadump /run/lizard-meta.dump
can't open metadata file
```

Then I came across the lizardfs-admin save-metadata command, and tried to use it from my lizardfs-slave... which gave me:
```
root@lizardfs-admin:/# lizardfs-admin save-metadata 10.210.162.72 9421
Admin password: 
Error: Authentication with the admin password failed: Operation not permitted
```

It seems I'm running a default /etc/mfs/mfsmaster.cfg on the master.  So the ADMIN_PASSWORD line is commented out... Is there a default admin password?

Any ideas?  Or should I reinstall and then rsync the files back onto the new chunkservers?

**Comment by psarna** [comment]
You almost succeeded, `lizardfs-admin save-metadata` is the way to go. There's no default password, just edit /etc/mfs/mfsmaster.cfg, add your ADMIN_PASSWORD and reload with `mfsmaster reload`.

Also, if you spawn a new shadow/metalogger somewhere, it will connect to master and download everything from it, so it will contain all the fresh metadata you need. That one doesn't involve changing anything in master configuration.

Finally, if your LizardFS installation is not under heavy load right now, you can simply take metadata.mfs file + all the changelog* files. It will all be sufficient to restore the installation. changelog.mfs file is human-readable, so you can read it (e.g. with `tail` or `less`) to see what was the last operation that got registered. There usually are several changelog* files, but changelog.mfs is always the most fresh one.

A side note: `mfsmetadump` is indeed a tool which reads metadata.mfs and translates its binary format to human-readable one, so it's not something you can use to extract fresh metadata image.

**Comment by 4Dolio** [comment]
Might be worth trying to remount=rw? Maybe overmount your /var/lib/mfs where metadata lives with a tmpfs?  Bring in a new shadow or restarting your broken shadow should get it back in sync, as mentioned this should be easiest way out for you.

**Comment by pauls1024** [comment]
Thanks for the quick responses and useful insights.

> /etc/mfs/mfsmaster.cfg, add your ADMIN_PASSWORD and reload with mfsmaster reload.
I cannot edit the config on master at present as it is readonly ...

I tried starting a shadow master but I'm seeing errors in the syslog.  I assumed this is because master cannot read its own /var/lib/mfs/metadata.mfs?
`Nov  2 15:19:55 lizardfs-slave mfsmaster[607]: some changes lost: [7182984-7291569], download metadata again
Nov  2 15:19:55 lizardfs-slave mfsmaster[607]: Waiting for master to produce up-to-date metadata image
Nov  2 15:19:55 lizardfs-slave mfsmaster[607]: Master temporarily refused to produce a new metadata image`


I can try a remount rw later on when it is quieter.

**Comment by 4Dolio** [comment]
Humm, that shadow error might be temporary, or maybe due to RO.. give the shadow a half hour to hour to try to sync up...  I think this should work, buy maybe it is dependent on the master writing to disk? Not sure.

If remount=rw fails, you might also try forcing mounting a tmpfs over the metadata folder, forcing for non-empty mountpoint if you must.

This is an edge case error that it would be nice if the master would warn about in cgi, and handle gracefully by ensuring a shadow/logger can still sync up based soely on metadata in ram only... feature request for heads up on master RO error in mfs.cgi server statuses tab?

**Comment by pauls1024** [comment]
Thanks for the assistance.  I was able to:
1) rsync off a current backup of all the files with their permissions and dates (rsync -avz)
2) start a new master with metadata from a 09:18 backup connected to the current chunk servers, which kept all the old files, but loses access to all files created since the filesystem went readonly.
3) rsync the backup back onto the lizardfs mount

So at the end of the day I have operation continuity: no loss of service and no files dropped.

Its just a wee bit frustrating to know that the valid metadata was all in memory on the running master, but there was no way I could down load it.

The OS crashed on that master when I tried to remount rw so no further testing is possible on that machine.

Thanks for your help.
